propolis/src/hw/nvme/cmds.rs

```diff
@@ -601,8 +607,26 @@ impl PrpIter<'_> {
                     // entry which should be present in PRP2
                     PrpNext::Prp2
                 } else {
-                    let list_off = (self.prp2 & PAGE_OFFSET as u64) / 8;
-                    PrpNext::List(self.prp2, list_off as u16)
+                    /*
+                     * The first PRP List entry (i.e. the first pointer to a memory page containing
+                     * additional PRP entries) that if present is contained in the PRP Entry 2 location
+                     * within the command, shall be Qword aligned and may also have a non-zero offset within
+                     * the memory page.
+                     */
+                    if (self.prp2 % 8) != 0 {
+                        return Err("PRP2 not Qword aligned!");
+                    }
+
+                    /*
+                     * If this entry is
+                     *
+                     * - not the first PRP entry in the command, or
+                     * - in the PRP List
+                     *
+                     * then the Offset portion of this field shall be cleared to 0h.
+                     */
+                    let idx = (self.prp2 & PAGE_OFFSET as u64) / 8;
+                    PrpNext::List(self.prp2 & (PAGE_MASK as u64), idx as u16)
```

Comment:
Is this constraint of the list base address proper?  I believe the base pointer to the PRP list can be offset within the page:

> Dependent on the command definition, the first PRP entry contained within the command may have a non-
zero  offset  within  the  memory  page.    The  first  PRP  List  entry  (i.e.  the  first  pointer  to  a  memory  page 
containing additional PRP entries) that if present is typically contained in the PRP Entry 2 location within 
the command, shall be Qword aligned and may also have a non-zero offset within the memory page.

Reply:
I'm not 100% sure.

The spec says what you quoted, and at the *same time* says (4.3, figure 10) that

> If this entry is not the first PRP entry in the command or in the PRP List then the Offset portion of this field shall be cleared to 0h.

I interpreted this to mean the following

- that prp2 can always come in with a non-zero offset
  - if not pointing to a list, then this is an offset into a page
  - if pointing to a list, that offset is interpreted as an index
- if prp2 is pointing to a list, then the offset should be zeroed.

I know that the tests I was running to validate this interpretation (https://fight-flash-fraud.readthedocs.io/en/latest/ on alpine, and separately an openbsd install) would report corruption (or panic) without this patch, and would work with it. That's not great justification but I believe it validates the interpretation.

Reply:
When the PRP2 value is referring to a PRP List, I believe the value itself does not count as a "PRP entry" (since it's not pointing to memory to be consumed in the operation).  Looking at the nvme emulation in both qemu and the C bhyve userspace, neither is masking PRP2 like that when using it to locate the PRP List.

Reply:
The reason this works is because we're constructing an enum:

```
                     let idx = (self.prp2 & PAGE_OFFSET as u64) / 8;
                     PrpNext::List(self.prp2 & (PAGE_MASK as u64), idx as u16)
```
that we later pull apart and use like this:
```
             PrpNext::List(base, idx) => {
                 assert!(idx <= PRP_LIST_MAX);
                 let entry_addr = base + (idx as u64) * 8;
```

If we don't mask off prp2, then base contains the offset, and idx adds the offset to that, pointing to an invalid memory location.

tl;dr my interpretation of the spec was wrong, but the fix was right :) I'll make a PR correcting the comments.